lib/Dialect/RTL/Ops.cpp

```diff
@@ -657,12 +657,53 @@ void AddOp::getCanonicalizationPatterns(OwningRewritePatternList &results,
         return success();
       }
 
+      // add(..., x, x) -> add(..., mul(x, 2))
+      if (inputs[size - 1] == inputs[size - 2]) {
+        SmallVector<Value, 2> mulInputs;
+
+        auto two = rewriter.create<ConstantOp>(
+            op.getLoc(), 2, op.getType().cast<IntegerType>());
+        mulInputs.push_back(inputs.back());
+        mulInputs.push_back(two);
+
+        auto mulOp = rewriter.create<rtl::MulOp>(op.getLoc(), mulInputs);
```

Comment:
Ok, sorry for the bad advice, but this is one step too far.  It is true that this will converge to the right solution (given that x*2 will turn into a shift) but there is no reason to ever intentionally generate non-canonical form.  Please change this to generate `x<<1` directly, thanks :-)

The reason to rely on recursive simplifications is if it provides a simplification to the compiler code itself.

Reply:
So to take a step back, you want the following:
1. add(x, x) -> shl(x, 1).
2. add(x, (shl(x, c) ) -> add( mul(x, (1 << c) + 1) )
3. add(x, mul(x, s)) -> add ( shl(x, log2(s + 1))
( The third isn't mentioned, but we'd need this to go back to shift lefts. Please let me know if I'm wrong here)

This seems to be what I had before, so maybe I'm mistaken here. The reason I went to this new form is because as you mentioned, the canonical pass of multiply should be taken care of in multiply which seemed logical. However, now you're saying we should avoid ever generating non-canonical form, so this requires more verbose code (above) in `add`'s pass. Let me know if I'm off target here.

Reply:
Close, but the third should be: `add(x, mul(x, s)) -> mul(x, s + 1)`.  The case when `s+1` is a power of two can be handled by recurisve simplifications.  We want to handle things like `add(x, mul(x, 13))` -> `mul(x, 14)`.